cnd/src/network/protocols/announce/handler.rs

```diff
@@ -0,0 +1,143 @@
+    }
+}
+
+/// Event produced by the `Handler`.
+#[derive(Debug)]
+pub enum HandlerEvent {
+    /// This event created when a confirmation message containing a `swap_id` is
+    /// received in response to an announce message containing a
+    /// `swap_digest`. The Event contains both the swap id and
+    /// the swap digest.
+    ReceivedConfirmation(Confirmed),
+
+    /// The event is created when a remote sends a `swap_digest`. The event
+    /// contains a reply substream for the receiver to send back the
+    /// `swap_id` that corresponds to the swap digest.
+    AwaitingConfirmation(ReplySubstream<NegotiatedSubstream>),
+
+    /// Failed to announce swap to peer.
+    Error(Error),
+}
+
+impl ProtocolsHandler for Handler {
+    type InEvent = OutboundConfig;
+    type OutEvent = HandlerEvent;
+    type Error = Error;
+    type InboundProtocol = InboundConfig;
+    type OutboundProtocol = OutboundConfig;
+    type OutboundOpenInfo = ();
+
+    fn listen_protocol(&self) -> SubstreamProtocol<Self::InboundProtocol> {
+        SubstreamProtocol::new(InboundConfig::default())
+    }
+
+    fn inject_fully_negotiated_inbound(
+        &mut self,
+        sender: <Self::InboundProtocol as InboundUpgrade<NegotiatedSubstream>>::Output,
+    ) {
+        self.events.push(HandlerEvent::AwaitingConfirmation(sender))
+    }
+
+    fn inject_fully_negotiated_outbound(
+        &mut self,
+        confirmed: <Self::OutboundProtocol as OutboundUpgrade<NegotiatedSubstream>>::Output,
+        _info: Self::OutboundOpenInfo,
+    ) {
+        self.events
+            .push(HandlerEvent::ReceivedConfirmation(confirmed));
+        self.keep_alive = KeepAlive::No;
+    }
+
+    fn inject_event(&mut self, event: Self::InEvent) {
+        self.keep_alive = KeepAlive::Yes;
+        self.dial_queue.push_front(event);
+    }
+
+    fn inject_dial_upgrade_error(
+        &mut self,
+        _info: Self::OutboundOpenInfo,
+        err: ProtocolsHandlerUpgrErr<
+            <Self::OutboundProtocol as OutboundUpgrade<NegotiatedSubstream>>::Error,
+        >,
+    ) {
+        self.events.push(HandlerEvent::Error(Error::Upgrade(err)));
+        self.keep_alive = KeepAlive::No;
+    }
+
+    fn connection_keep_alive(&self) -> KeepAlive {
+        self.keep_alive
+    }
+
+    fn poll(
+        &mut self,
+        _: &mut Context<'_>,
+    ) -> Poll<
+        ProtocolsHandlerEvent<
+            Self::OutboundProtocol,
+            Self::OutboundOpenInfo,
+            HandlerEvent,
+            Self::Error,
+        >,
+    > {
+        if !self.events.is_empty() {
+            let event = self.events.remove(0);
+            if let HandlerEvent::Error(err) = event {
+                return Poll::Ready(ProtocolsHandlerEvent::Close(err));
+            };
+            return Poll::Ready(ProtocolsHandlerEvent::Custom(event));
+        }
+
+        if !self.dial_queue.is_empty() {
+            let upgrade = self.dial_queue.remove(0).unwrap(); // Cannot fail, queue is not empty.
```

Comment:
```
error[E0599]: no method named `pop` found for struct `std::collections::VecDeque<network::protocols::announce::protocol::OutboundConfig>` in the current scope
   --> cnd/src/network/protocols/announce/handler.rs:125:46
    |
125 |         if let Some(event) = self.dial_queue.pop() {
    |                                              ^^^ method not found in `std::collections::VecDeque<network::protocols::announce::protocol::OutboundConfig>`
```

Reply:
Ah if it is a `VecDeque` the fn is called `pop_front` or something, look it up in the docs :)

Reply:
yeh changed it to pop_front with if let Some(..)

Reply:
Ok, this one was never resolved.  I had a quick hack, and it looks like this is a bigger issue than just using the correct function.  We (mis)use a `Vec` as a FILO queue for `events` and we (mis)use a `VecDequeue` as a FIFO queue.  We should decide what behaviour we _actually_ want here and also we should use the data structures correctly.  Since bors is running I'll add a ticket for this.

Reply:
Tracking issue: https://github.com/comit-network/comit-rs/issues/2306